File: LeetCode/Easy/0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        count = defaultdict(int)
        max_val = 0
        def DFS(node:Optional[TreeNode]):
            if not node:
                return

            count[node.val] += 1
            DFS(node.left)
            DFS(node.right)
        
        DFS(root)

        for v in count.values():
            if v > max_val:
                max_val = v

        return [k for k, v in count.items() if v == max_val]

File: LeetCode/Easy/0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<int, int> count;
    vector<int> findMode(TreeNode* root) {
        DFS(root);
        int max_val = 0;
        vector<int> ret;

        for(auto elem: count){
            if (elem.second > max_val) {
                ret.clear();
                ret.push_back(elem.first);
                max_val = elem.second;
            } else if (elem.second == max_val) ret.push_back(elem.first);
        }
        return ret;
    }

    void DFS(TreeNode* node){
        if (!node) return;
        if (count.find(node->val) == count.end()){
            count[node->val] = 1;
        } else ++count[node->val];
        DFS(node->left);
        DFS(node->right);
    }
};